### PR TITLE
Redux: Ensures invalid connections are not re-used

### PR DESCRIPTION
**Pull request by hsume2**
Turns out my test didn't completely exercise the code the same way `rake
db:reset` exercises the code. Here's why the test was inadequate:

1. It runs `TestNotThere.establish_connection` followed by
`TestNotThere.connection`, which properly removes the invalid connection
to `arhp_test_no_create` from the connection pool.
2. Then it runs `Test1.establish_connection(:test_host_1_db_1)`
(database: `arhp_test_1`).
3. This invokes
https://github.com/rails/rails/blob/v3.2.21/activerecord/lib/active_record/connection_adapters/abstract/connection_specification.rb#L136
4. Which invokes
https://github.com/rails/rails/blob/v3.2.21/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb#L429
5. Which invokes `self.__getobj__` https://github.com/zendesk/active_record_host_pool/blob/master/lib/active_record_host_pool/pool_proxy.rb#L21
6. Which re-instantiates the connection pool with the original spec: https://github.com/zendesk/active_record_host_pool/blob/master/lib/active_record_host_pool/pool_proxy.rb#L104
7. Which stores an invalid connection to `arhp_test_1`.
8. And as expected, when it runs `Test1.connection`, it gets an `Unknown
database 'arhp_test_1'` exception.

And here is where the use case breaks down:

1. It runs `TestNotThere.establish_connection` followed by
`TestNotThere.connection`, which properly removes the invalid connection
to `arhp_test_no_create` from the connection pool.
2. Then it runs `TestNotThere.establish_connection(:test_host_1_db_1)`
(database: `arhp_test_1`).
3. This invokes
https://github.com/rails/rails/blob/v3.2.21/activerecord/lib/active_record/connection_adapters/abstract/connection_specification.rb#L136
4. Which invokes
https://github.com/rails/rails/blob/v3.2.21/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb#L429
5. Which invokes `self.__getobj__` https://github.com/zendesk/active_record_host_pool/blob/master/lib/active_record_host_pool/pool_proxy.rb#L21
6. Which re-instantiates the connection pool with the original spec: https://github.com/zendesk/active_record_host_pool/blob/master/lib/active_record_host_pool/pool_proxy.rb#L104
7. Which stores an invalid connection to `arhp_test_no_create` (since the @spec is cached from 1)
8. And when it runs `TestNotThere.connection`, it gets an `Unknown
database 'arhp_test_no_create'` exception.

In conclusion, this PR will get us past the problem. Though, in all
seriousness, if there is an invalid connection and we manage to
hit an undefined method in `PoolProxy`, we will store that invalid
connection. It seems a longer term fix would be to check for connection
validity prior to storing/retrieving a connection.

### CODE REVIEW

**Comment by hsume2** [comment]
@osheroff so close, yet so far. This one fixes it and I've validated it all the way through Classic `rake db:reset` this time.

**Comment by hsume2** [comment]
@osheroff maybe you missed this one, this is the proper fix

**Comment by osheroff** [comment]
give it a go! :+1:

**Comment by hsume2** [comment]
@osheroff, could you push v0.8.3 or add me as a maintainer in rubygems? Thanks!
